File: lib/llm/claude_cli.rb
# frozen_string_literal: true

require_relative "agent"

module LLM
  class Agent
    # The local `claude` CLI, run headless with -p. This is the original
    # Tui::Claude spawn logic moved onto the Agent protocol with zero behavior
    # change. `claude -p --output-format text` streams the assistant's text the
    # same way whether or not we want a transcript, so the `stream:` hint is a
    # no-op here — it only matters for harnesses with distinct one-shot vs.
    # transcript modes (see Hermes).
    class ClaudeCli < Agent
      def self.default_command = "claude"

      def command(prompt, model:, stream: true)
        # --dangerously-skip-permissions: a headless run can't answer permission
        # prompts, and the whole point is letting the agent edit gtd.org freely.
        args = [@command, "-p", prompt, "--model", model,
                "--output-format", "text", "--dangerously-skip-permissions"]
        args += ["--append-system-prompt", @system] if @system
        args
      end
    end
  end
end

File: internal/llm/claudecli.go
package llm

// ClaudeCLI is the local `claude` CLI, run headless with -p.
//
// `claude -p --output-format text` streams the assistant's text the same way
// whether or not a transcript is wanted, so the `stream` hint is a no-op here —
// it only matters for harnesses with distinct one-shot and transcript modes
// (see Hermes).
type ClaudeCLI struct{}

func (ClaudeCLI) DefaultBinary() string { return "claude" }

// Reachable: nothing beyond the PATH probe. An installed `claude` is a usable
// `claude`; it reaches its model over the network on its own.
func (ClaudeCLI) Reachable() bool { return true }

func (ClaudeCLI) Argv(binary, system, prompt, model string, _ bool) []string {
	// --dangerously-skip-permissions: a headless run cannot answer permission
	// prompts, and the whole point is letting the agent edit tasks.jsonl freely.
	argv := []string{binary, "-p", prompt, "--model", model,
		"--output-format", "text", "--dangerously-skip-permissions"}
	if system != "" {
		argv = append(argv, "--append-system-prompt", system)
	}
	return argv
}
